raitools/raitools/_managers/counterfactual_manager.py

```diff
@@ -2,33 +2,204 @@
+                    ' It should be the class value')
+
+        if self._task_type == ModelTask.REGRESSION:
+            if new_counterfactual_config.desired_range is None:
+                raise UserConfigValidationException(
+                    'The desired_range should not be None'
+                    ' for regression scenarios')
+
+        duplicate_counterfactual_config_found = False
+        for counterfactual_config in self._counterfactual_config_list:
+            if counterfactual_config == new_counterfactual_config:
+                duplicate_counterfactual_config_found = True
+                break
+
+        if duplicate_counterfactual_config_found:
+            raise DuplicateCounterfactualConfig(
+                'Duplicate counterfactual configuration detected')
+        else:
+            self._counterfactual_config_list.append(new_counterfactual_config)
 
     def add(self,
-            continuous_features=None,
-            outcome_name=None,
-            total_CFs=None,
+            continuous_features,
+            total_CFs,
+            method='random',
+            desired_class=None,
             desired_range=None,
-            desired_class="opposite",
             permitted_range=None,
-            features_to_vary=None,
-            method=None):
-        raise NotImplementedError(
-            "Add not implemented for CounterfactualManager")
+            features_to_vary=None):
+        """Add a counterfactual generation configuration to be computed later.
+        """
+
+        counterfactual_config = _CounterfactualConfig(
+            method=method,
+            continuous_features=continuous_features,
+            total_CFs=total_CFs,
+            desired_class=desired_class,
+            desired_range=desired_range,
+            permitted_range=permitted_range,
+            features_to_vary=features_to_vary)
+
+        self._add_counterfactual_config(counterfactual_config)
 
     def compute(self):
-        pass
+        """Computes the counterfactual examples by running the counterfactual
+           configuration."""
+        for cf_config in self._counterfactual_config_list:
+            if not cf_config.is_computed:
+                cf_config.is_computed = True
+                try:
+                    dice_explainer = self._create_diceml_explainer(
+                        method=cf_config.method,
+                        continuous_features=cf_config.continuous_features)
+
+                    X_test = self._test.drop([self._target_column], axis=1)
+
+                    counterfactual_obj = \
+                        dice_explainer.generate_counterfactuals(
+                            X_test, total_CFs=cf_config.total_CFs,
+                            desired_class=cf_config.desired_class,
+                            desired_range=cf_config.desired_range)
+
+                    cf_config.counterfactual_obj = \
+                        counterfactual_obj
+                except Exception as e:
```

Comment:
When do we expect this to fail? Should we be logging the error message to the module logger if there's a failure?

Reply:
Yes, this could fail if there is a bug in dice-ml. The failure reason is captured inside the failure reason. Not sure if there is a need to log it as well.